src/types/routeMethods.ts

```diff
@@ -58,17 +58,23 @@ export type ExtractRouteMethodParams<T> = T extends RouteMethod<infer Params>
 export type ExtractParamsFromPath<
   TPath extends Route['path']
 > = TPath extends Path
-  ? ExtractParamsFromPathString<TPath['path'], TPath['params']>
+  ? ExtractParamsFromPathString<UnifyParamEnds<TPath['path']>, TPath['params']>
   : TPath extends string
-    ? ExtractParamsFromPathString<TPath>
+    ? ExtractParamsFromPathString<UnifyParamEnds<TPath>>
     : never
 
+type ParamEnd = '/'
+
+type UnifyParamEnds<
+  TPath extends string
+> = ReplaceAll<ReplaceAll<TPath, '-', ParamEnd>, '_', ParamEnd>
```

Comment:
rather than nesting `ReplaceAll`, could we use a union?

```ts
ReplaceAll<TPath, '-' | '_', ParamEnd>
```

Reply:
Every time you have a union the type will produce a union. And we want to replace each character individually so the output type is a single template literal.